### PR TITLE
quodlibet: fix tests

### LINKED ISSUE #53938: quodlibet-full fails to install on nixos-unstable
## Issue description

`quodlibet-full` fails to install on `nixos-unstable`. It worked before the recent update. It fails early on in the check phase.

```
running install tests
============================= test session starts ==============================
platform linux -- Python 3.7.2, pytest-3.9.3, py-1.7.0, pluggy-0.8.0
rootdir: /build/quodlibet-4.2.1, inifile: setup.cfg
plugins: xdist-1.24.1, forked-0.2
collected 3035 items / 1 skipped

tests/test___init__.py .....                                             [  0%]
tests/test_appdata_files.py ssss                                         [  0%]
tests/test_browsers___init__.py .........                                [  0%]
tests/test_browsers__base.py ........................................... [  2%]
.......................................................
builder for '/nix/store/rxl4kyd2frj3an3ywwzgmf8v9hg9p1mb-quodlibet-full-4.2.1.drv' failed with exit code 139
```

I'm guessing the fact that I have to build it from source at all is because it also fails everywhere else, so there's no cached build to download.

### Steps to reproduce

Try to install `quodlibet-full` on `nixos-unstable`

## Technical details
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.13, NixOS, 19.03.git.626233e (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2`
 - channels(root): `"nixos-19.03pre166308.626233eee6e"`
 - nixpkgs: `/pkgs/nixpkgs`

### PR DESCRIPTION
**Pull request by 101100**
###### Motivation for this change

Attempting to fix Quodlibet by building on the comments in #61219.
Fixes #53938

###### Things done

Steps performed:

- excluded the quality tests and removed extra inputs for those tests
- added `$XDG_ICON_DIRS` to `XDG_DATA_DIRS` as in done in `wrap-gapps-hook.sh`
- added `gdk-pixbuf` to `checkInputs`

I'm not sure on the syntax I used for adding `gdk-pixbuf`.  It feels awkward; perhaps there is a better way to make that change?

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by 101100** [comment]
I ran `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"` but I can't tell if the output is a "pass":

```
$ git -c fetch.prune=false fetch --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0
remote: Enumerating objects: 10, done.
remote: Counting objects: 100% (10/10), done.
remote: Compressing objects: 100% (5/5), done.
remote: Total 12 (delta 5), reused 5 (delta 5), pack-reused 2
Unpacking objects: 100% (12/12), done.
From https://github.com/NixOS/nixpkgs
   3c2e9fa53d2..6d24724adc1  master     -> refs/nixpkgs-review/0
$ git worktree add /home/jasonpheard/.cache/nixpkgs-review/rev-f1ead65a6eef5033b34bd199f69348e07ad353cb-dirty/nixpkgs 6d24724adc1be6a470af80424b9aef3d00288d25
Preparing worktree (detached HEAD 6d24724adc1)
Updating files: 100% (20762/20762), done.
HEAD is now at 6d24724adc1 Merge pull request #77728 from dywedir/skim
$ nix-env -f /home/jasonpheard/.cache/nixpkgs-review/rev-f1ead65a6eef5033b34bd199f69348e07ad353cb-dirty/nixpkgs -qaP --xml --out-path --show-trace
No diff detected, stopping review...
$ git worktree prune
```

**Comment by 101100** [comment]
I tried running `nix path-info -S nixpkgs.quodlibet` but it seems to be using the system `nixpkgs` as it says that the package is broken.  How do I run it on the git repository?

**Comment by sauyon** [comment]
> I ran `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"` but I can't tell if the output is a "pass":

I don't think there are packages that depend on quodlibet, so this shouldn't be an issue.

> I tried running `nix path-info -S nixpkgs.quodlibet` but it seems to be using the system `nixpkgs` as it says that the package is broken. How do I run it on the git repository?

I think you're missing `-I <path/to/nixpkgs>`. Let me try this out.

**Comment by sauyon** [comment]
This still doesn't build for me, because `operon` tests fail. I suspect this has something to do with it trying to write to non-writable paths, but it's inconsequential enough that I think just adding `--ignore=tests/test_operon.py` is fine.

FWIW you should run `nix path-info -S <path to your built derivation>`.

**Comment by alyssais** [comment]
Please format your commit messages according to CONTRIBUTING.md.

**Comment by 101100** [comment]
@alyssais Done.

@sauyon What environment/version are you using to get the failed tests?

Also, is there an easy way to automatically build all of the possible iterations of quodlibet?  I manually built `quodlibet`, `quodlibet-full`, `quodlibet-xine`, `quodlibet-xine-full` and `quodlibet-without-gst-plugins` but that seems like a very manual process.

**Comment by sauyon** [comment]
Arch Linux, x84_64.

**Comment by 101100** [comment]
@sauyon I wasn't able to get the build to fail with Arch in docker, but I can just disable that test if that is the only way it works for you.  Is that what you recommend?

**Comment by sauyon** [comment]
Give me a little bit to see if I can fix it. Not sure what the official stance is on these things, but it's just a test and honestly I'd prefer to have a package that I can install that may have some broken parts than nothing :P.

File: pkgs/applications/audio/quodlibet/default.nix [hunk 1]
```diff
@@ -35,11 +35,11 @@ python3.pkgs.buildPythonApplication rec {
 
   checkPhase = ''
     runHook preCheck
-    env XDG_DATA_DIRS="$out/share:${gtk3}/share/gsettings-schemas/${gtk3.name}:$XDG_DATA_DIRS" \
+    env XDG_DATA_DIRS="$out/share:${gtk3}/share/gsettings-schemas/${gtk3.name}:$XDG_ICON_DIRS:$XDG_DATA_DIRS" \
       HOME=$(mktemp -d) \
       xvfb-run -s '-screen 0 800x600x24' dbus-run-session \
         --config-file=${dbus.daemon}/share/dbus-1/session.conf \
-        py.test${stdenv.lib.optionalString (xineBackend || !withGstPlugins) " --ignore=tests/plugin/test_replaygain.py"}
+        py.test ${stdenv.lib.optionalString (xineBackend || !withGstPlugins) " --ignore=tests/plugin/test_replaygain.py"} --ignore=tests/quality
```
Comment (jtojnar) [inline pkgs/applications/audio/quodlibet/default.nix]: Does `--no-quality` suggested in the other PR not work?

Comment (101100) [reply pkgs/applications/audio/quodlibet/default.nix]: No, that is part of the testing script from `quodlibet` that this is not running.  It seems that this build script runs the tests directly instead and errored with `--no-quality`.

Here is where it is defined in quodlibet:

https://github.com/quodlibet/quodlibet/blob/30293621f466cbc0076ed2db703702ef0234a853/quodlibet/gdist/tests.py#L42

**Comment by sauyon** [comment]
My initial stab resulted in not much success and I'm likely to be busy over MLK weekend, so I'm happy with this with the additional tests ignored.

**Comment by 101100** [comment]
@sauyon I've excluded the test you indicated.

**Review by sauyon — APPROVED** [review]
LGTM, thanks!

**Review by sauyon — APPROVED** [review]
It seems the :robot: has decided that my review was not sufficient.

**Comment by jtojnar** [comment]
@GrahamcOfBorg build quodlibet quodlibet-full quodlibet-xine quodlibet-xine-full quodlibet-without-gst-plugins

@sauyon could you share the error from openon test failing?

**Comment by sauyon** [comment]
Here's the error output:

```
=================================== FAILURES ===================================
___________________________ TOperonEdit.test_dry_run ___________________________

self = <tests.test_operon.TOperonEdit testMethod=test_dry_run>

    def test_dry_run(self):
        if os.name == "nt" or sys.platform == "darwin":
            return

        realitems = lambda s: [(k, s[k]) for k in s.realkeys()]

        os.environ["VISUAL"] = "truncate -s 0"
        old_items = realitems(self.s)
        os.utime(self.f, (42, 42))
>       e = self.check_true(["edit", "--dry-run", self.f], False, True)[1]

tests/test_operon.py:369:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/test_operon.py:55: in check_true
    return self._check(args, True, so, se, **kwargs)
tests/test_operon.py:64: in _check
    self.failUnlessEqual(s == 0, success, msg=repr((s, o, e)))
tests/__init__.py:40: in assertEqual
    super().assertEqual(second, first, msg)
E   AssertionError: True != False : (1, '', 'edit: No changes detected\n')
_________________________ TOperonEdit.test_remove_all __________________________

self = <tests.test_operon.TOperonEdit testMethod=test_remove_all>

    def test_remove_all(self):
        if os.name == "nt" or sys.platform == "darwin":
            return

        os.environ["VISUAL"] = "truncate -s 0"
        os.utime(self.f, (42, 42))
>       self.check_true(["edit", self.f], False, False)

tests/test_operon.py:385:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/test_operon.py:55: in check_true
    return self._check(args, True, so, se, **kwargs)
tests/test_operon.py:64: in _check
    self.failUnlessEqual(s == 0, success, msg=repr((s, o, e)))
tests/__init__.py:40: in assertEqual
    super().assertEqual(second, first, msg)
E   AssertionError: True != False : (1, '', 'edit: No changes detected\n')
```

My guess is it's some sort of permissions error, but I haven't had the time to look into too closely.

File: pkgs/applications/audio/quodlibet/default.nix [hunk 2]
```diff
@@ -35,11 +35,11 @@ python3.pkgs.buildPythonApplication rec {
```
Comment (jtojnar) [inline pkgs/applications/audio/quodlibet/default.nix]: I would suggest to refactor it like this, both to make it cleaner and to be able to comment the ignores:

```suggestion
  pytestFlags = stdenv.lib.optionals (xineBackend || !withGstPlugins) [
    "--ignore=tests/plugin/test_replaygain.py"
  ] ++ [
    # upstream does actually not enforce source code linting
    "--ignore=tests/quality"
    # build failure on Arch Linux
    # https://github.com/NixOS/nixpkgs/pull/77796#issuecomment-575841355
    "--ignore=tests/test_operon.py"
  ];

```

File: pkgs/applications/audio/quodlibet/default.nix [hunk 3]
```diff
@@ -35,11 +35,11 @@ python3.pkgs.buildPythonApplication rec {
 
   checkPhase = ''
     runHook preCheck
-    env XDG_DATA_DIRS="$out/share:${gtk3}/share/gsettings-schemas/${gtk3.name}:$XDG_DATA_DIRS" \
+    env XDG_DATA_DIRS="$out/share:${gtk3}/share/gsettings-schemas/${gtk3.name}:$XDG_ICON_DIRS:$XDG_DATA_DIRS" \
       HOME=$(mktemp -d) \
       xvfb-run -s '-screen 0 800x600x24' dbus-run-session \
         --config-file=${dbus.daemon}/share/dbus-1/session.conf \
-        py.test${stdenv.lib.optionalString (xineBackend || !withGstPlugins) " --ignore=tests/plugin/test_replaygain.py"}
+        py.test ${stdenv.lib.optionalString (xineBackend || !withGstPlugins) " --ignore=tests/plugin/test_replaygain.py"} --ignore=tests/quality --ignore=tests/test_operon.py
```
Comment (jtojnar) [inline pkgs/applications/audio/quodlibet/default.nix]:
```suggestion
        py.test $pytestFlags
```

to go with the above.

```suggestion
        py.test "''$pytestFlags[@]}"
```

would be better but we cannot use that without #72074.

File: pkgs/applications/audio/quodlibet/default.nix (hunk 2 above)
Comment (101100) [reply pkgs/applications/audio/quodlibet/default.nix]: Done; I've pushed another commit.  Is the practice to keep it down to a single commit, though?

File: pkgs/applications/audio/quodlibet/default.nix (hunk 3 above)
Comment (101100) [reply pkgs/applications/audio/quodlibet/default.nix]: ...and done.

**Comment by 101100** [comment]
@GrahamcOfBorg build quodlibet quodlibet-full quodlibet-xine quodlibet-xine-full quodlibet-without-gst-plugins

**Comment by 101100** [comment]
@jtojnar @sauyon All comments have been addressed; ready for another round of reviews.

**Review by sauyon — APPROVED** [review]
Sorry, this looks fine to me.

**Comment by 101100** [comment]
@jtojnar How can I get this merged?

**Comment by jtojnar** [comment]
Thanks.

**Comment by 101100** [comment]
Thanks!  Could I merge this change into 19.09 too since it is broken there?

**Pull request merged**